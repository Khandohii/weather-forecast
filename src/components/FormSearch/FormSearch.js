import './FormSearch.scss';
import React, { useEffect, useState} from 'react';
import { Transition } from 'react-transition-group';
import useOpenStreetMapService from '../../services/openStreetMapService';
import ErrorMessage from '../ErrorMessage/ErrorMessage';

const FormSearch = (props) => {
    const [showList, setShowList] = useState(false);
    const [searchData, setSearchData] = useState('');
    const [shouldSearch, setShouldSearch] = useState(false);
    const [searchTimeout, setSearchTimeout] = useState(null);

    const handleFocus = () => {
        setShowList(true);
    };

    const handleBlur = () => {
        setShowList(false);
    };

    const onInputChange = (event) => {
        setSearchData(event.target.value);
        clearTimeout(searchTimeout);
        setSearchTimeout(setTimeout(() => setShouldSearch(true), 500));
    };

    useEffect(() => {
        return () => clearTimeout(searchTimeout);
    }, [searchTimeout]);

    return (
        <div className={'form-search ' + props.className}>
            <form action="" method="get">
                <input 
                    name='city'
                    type="text" 
                    placeholder='Search City' 
                    onFocus={handleFocus}
                    onBlur={handleBlur}
                    onChange={onInputChange}
                />
            </form>

            <ResultList show={showList} searchData={searchData} shouldSearch={shouldSearch} setCityFunc={props.setCityFunc} setCoordsGlobal={props.setCoordsGlobal} setShouldSearch={setShouldSearch} />
        </div>
    )
}

const ResultList = (props) => {
    const [cityList, setCityList] = useState([]);
    const {loading, error, getLocationData} = useOpenStreetMapService();

    const duration = 300;
    
    const defaultStyle = {
      transition: `opacity ${duration}ms ease-in-out`,
      opacity: 0,
      visibility: 'hidden'
    }
    
    const transitionStyles = {
      entering: { opacity: 1, visibility: 'visible' },
      entered:  { opacity: 1, visibility: 'visible' },
      exiting:  { opacity: 0, visibility: 'visible' },
      exited:  { opacity: 0 },
    };

    useEffect(() => {
        if (props.shouldSearch && props.searchData.length > 2) {
            updateCitiesList(props.searchData);
            props.setShouldSearch(false);
        }
    }, [props.shouldSearch, props.searchData]);

    const updateCitiesList = (city) => {
        getLocationData(city).then((res) => {
            onCitiesListLoaded(res)
        })
    }

    const onCitiesListLoaded = (data) => {
        setCityList(data);
    }

    function renderItems(items) {
        const cities = items.map((item, i) => {    
            return(
                <li key={item.id}>
                    <button 
                        onClick={() => {
                            props.setCoordsGlobal(item.coords)
                            props.setCityFunc(item.name)
                        }}
                    >
                        {item.display_name}
                    </button>
                </li>
            )
        })

        if (props.searchData.length < 3) {
            return "Input minumum 3 symbols"
        }

        if (items.length < 1) {
            return "City not found"
        }

        return (
            <ul>
                {cities}
            </ul>
        );
    }

    const items = renderItems(cityList);

    const errorMessage = error ? <ErrorMessage/> : null;
    const spinner = loading ? "Loading..." : null;

    return(
        <Transition 
            in={props.show} 
            timeout={duration}
            unmountOnExit
        >
            {state => (
                <div 
                    className='form-search__results'
                    style={{
                        ...defaultStyle,
                        ...transitionStyles[state]
                    }}
                >
                    {errorMessage}
                    {spinner || items}
                </div>
            )}
        </Transition>
    )
}

export default FormSearch;
