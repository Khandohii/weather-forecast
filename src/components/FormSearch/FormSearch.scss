@import '../../style/variables.scss';

.form-search{
    position: relative;
    z-index: 10;

    input{
        display: block;
        width: 100%;
        font-family: $font-family;
        font-size: 1.2em;
        padding: 0 25px;
        padding-left: 60px;
        height: 60px;
        border-radius: 10px;
        border: none;
        background: rgba($color: $bg-blue, $alpha: .4) url('../../resources/icons/search-icon.svg') 25px 50% no-repeat;
        color: $main-color;
    }

    &__results{
        position: absolute;
        left: 0;
        top: calc(100% + 15px);
        background: $main-color;
        color: $bg-blue;
        padding: 15px;
        border-radius: 10px;
        width: 100%;
        box-shadow: 0px 10px 10px 0px rgba(0, 0, 0, 0.2);

        li{
            list-style: none;

            & + li{
                margin-top: 5px;
            }

            button{
                padding: 10px 15px;
                transition: $transition;
                cursor: pointer;
                border-radius: 10px;
                display: block;
                width: 100%;
                border: none;
                background-color: transparent;
                text-align: left;
                font-family: $font-family;
                font-size: 1.05em;

                &:hover{
                    background: rgba($color: $bg-blue, $alpha: .6);
                    color: $main-color;
                }
            }
        }
    }
}