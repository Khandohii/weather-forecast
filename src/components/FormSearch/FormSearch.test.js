import renderer from 'react-test-renderer';
import React from 'react';
import FormSearch from './FormSearch.js';

it('renders FormSearch component correctly', () => {
    const component = renderer.create(
        <FormSearch setCityFunc={() => {}} setCoordsGlobal={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})