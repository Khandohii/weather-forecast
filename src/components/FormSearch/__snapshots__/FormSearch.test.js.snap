// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FormSearch component correctly 1`] = `
<div
  className="form-search undefined"
>
  <form
    action=""
    method="get"
  >
    <input
      name="city"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Find city"
      type="text"
    />
  </form>
</div>
`;
